native/src/fairseq2n/data/concat_data_source.cc

```diff
@@ -13,11 +13,11 @@ namespace fairseq2n::detail {
 concat_data_source::concat_data_source(std::vector<data_pipeline> &&pipelines) noexcept
   : pipelines_(std::move(pipelines))
 {
-    is_infinite_ = std::any_of(
-        pipelines_.begin(), pipelines_.end(), [](const data_pipeline &p)
+    finitude_type_ = std::max_element(
```

Comment:
Nit: For better readability, I would separate `std::max_element()` and `->get_finitude_type()` statements.